c10/util/Exception.cpp

```diff
@@ -147,25 +147,53 @@ thread_local WarningHandler* ThreadWarningHandler::warning_handler_ = nullptr;
 
 } // namespace
 
-void warn(
+Warning::Warning(
     const SourceLocation& source_location,
     const std::string& msg,
-    const bool verbatim) {
-  ThreadWarningHandler::get_handler()->process(source_location, msg, verbatim);
-}
+    const bool verbatim)
+    : source_location_(source_location),
+      msg_(std::move(msg)),
+      verbatim_(verbatim) {}
 
-void warn(
+Warning::Warning(
     SourceLocation source_location,
     detail::CompileTimeEmptyString msg,
-    const bool verbatim) {
-  warn(source_location, "", verbatim);
-}
+    const bool verbatim)
+    : Warning(std::move(source_location), "", verbatim) {}
 
-void warn(
+Warning::Warning(
     SourceLocation source_location,
     const char* msg,
-    const bool verbatim) {
-  ThreadWarningHandler::get_handler()->process(source_location, msg, verbatim);
+    const bool verbatim)
+    : source_location_(std::move(source_location)),
+      msg_(std::string(msg)),
+      verbatim_(verbatim) {}
+
+Warning::~Warning() {}
+
+const SourceLocation& Warning::source_location() const {
+  return source_location_;
+}
+
+const std::string& Warning::msg() const {
+  return msg_;
+}
+
+bool Warning::verbatim() const {
+  return verbatim_;
+}
+
+void Warning::warn() const {
+  TORCH_INTERNAL_ASSERT(
+      false, "Cannot call `Warning::warn()`, you must use a subclass of `Warning`");
+}
+
+// TODO: I don't like that subclasses currently override this method just
+// so they can do `make_shared<subclass_type>`. Is there another way to make
+// a shared_ptr of the proper type without needing to override this method for
+// each subclass?
+void UserWarning::warn() const {
+  ThreadWarningHandler::get_handler()->process(std::make_shared<UserWarning>(*this));
```

Comment:
Another option is to make `warn()` a factory template. It looks like the macros are used to instantiate warning objects and `.warn()` is called immediately. So the factory becomes responsible for creating the object and immediately passing to the handler. 

```c++
template<typename WarningType, typename... Args>
void warn(Args&&... args) { 
   ThreadWarningHandler::get_handler()->process(std::make_shared<WarningType>(std::forward<Args>(args)...));
}
```

Reply:
That's a good solution, thanks!

Reply:
Actually, I'm not sure this works the way we need it to. This template function has to be called like `warn<WarningType>(...);`, with the warning subclass explicitly specified. That's a problem in `torch/csrc/autograd/utils/warnings.cpp`--when warnings are replayed from the buffer, they could be warnings of any subclass type. So I think we have to stick with using the `Warning::warn()`/`UserWarning::warn()` methods, since they don't need the subclass to be specified explicitly

Reply:
When replaying the warnings from the buffer you do not need to create the shared ptrs, as you already stored them. You can simply `process()` them with the appropriate handler.

Reply:
For what is worth, if you have an a-priori limited set of warnings and you want to store them and process them in a uniform way, you may want to use `c10::variant` together with the `visit` method. This would save you from using pointers at all.

Reply:
Note that since in C++14 we don't have variadic lambdas, you'd need to implement this via your own functor. Now, if you don't want to pollute the global namespace, remember that you can define the struct to be local to the function you are using it in. This is as close as you get to variadic lambdas in C++14, I guess

Reply:
@amjames, I see, you're right. However, today I tried again to implement this template `warn` function, and I've been having some trouble getting it exactly right--I've gotten various build errors and undefined symbols at runtime. Since the template has to go in a header file, every time I change it I have to rebuild for 30+ minutes, which is a pain. I'll keep looking into it, but maybe this should be done in a follow-up PR

Reply:
Okay, I got the template function factory `c10::warn()` working and removed `c10::Warning::warn()`. I also added two more `c10::warn()` overloads that take in a `c10::Warning` by reference or a smart pointer to a `c10::Warning`--the second one helped with file scoping so that I didn't need to move the `ThreadWarningHandler` definition from `c10/utils/Exceptions.cpp` into the header file

Reply:
@Lezcano, I'm not sure if I fully understand your suggestion yet, I'm not familiar with `c10::variant`, so let me know if my interpretation is right.

It seems like part of your suggestion is that instead of, for instance, `void PyWarningHandler::InternalHandler::process(std::shared_ptr<c10::Warning> warning)` (in the `torch/csrc/Exceptions.cpp` file), we could do `void PyWarningHandler::InternalHandler::process(c10::variant<c10::UserWarning, ...> warning_variant)`. Then inside that function we would call `c10::visit([&](auto&& warning) {...}, warning_variant);`, and `warning` would have the proper subclass type inside the `{...}` scope there.

It's worth mentioning that as we add more warning subclasses, we'd have to keep adding them to `c10::variant<c10::UserWarning, ...>`. But we can just make a typedef for it like `typedef c10::variant<c10::UserWarning, ...> warning_variant_t` to be reused wherever needed

Reply:
That's exactly on point. The only thing is that `auto&&` is not C++14, but yep. And yes, we should make a `using variant_type = ...`

Reply:
Alright, I've implemented the variant type and it seems to be working. One possible issue with it is that the `process(c10::warning_variant_t)` functions are passing the variant by value. Looks like passing a variant by reference like `process(c10::warning_variant_t&)` is not allowed. However, I could create another variant type that holds references to warnings, like `typedef c10::variant<std::reference_wrapper<UserWarning>> warning_ref_variant_t;` to use in these situations. Is it worth the extra complexity? Or is there a better way to do it?

Reply:
I would expect that you can pass it by const reference. A variant is nothing but a glorified union with extra type-safety, so it behaves just as a normal type. What error do you get?

Reply:
Oh, you're right. The error I was getting was actually because I tried to use `vector<const c10::warning_variant_t&>`. I guess it's alright to just use `vector<c10::warning_variant_t>`

Reply:
yeah, no, a vector cannot store references.